starknet/src/utils/types.cairo

```diff
@@ -115,33 +108,74 @@ impl ProposalStatusIntoU8 of Into<ProposalStatus, u8> {
     }
 }
 
-impl ProposalStatusPartialEq of PartialEq<ProposalStatus> {
-    fn eq(lhs: ProposalStatus, rhs: ProposalStatus) -> bool {
-        ProposalStatusIntoU8::into(lhs) == rhs.into()
+impl LegacyHashChoice of LegacyHash<Choice> {
+    fn hash(state: felt252, value: Choice) -> felt252 {
+        LegacyHash::hash(state, ChoiceIntoU8::into(value))
     }
+}
 
-    fn ne(lhs: ProposalStatus, rhs: ProposalStatus) -> bool {
-        ProposalStatusIntoU8::into(lhs) != rhs.into()
-    }
+#[derive(Option, Clone, Drop, Serde, StorageAccess)]
+struct Strategy {
+    address: ContractAddress,
+    params: Array<felt252>,
 }
 
-impl FinalizationStatusPartialEq of PartialEq<FinalizationStatus> {
-    fn eq(lhs: FinalizationStatus, rhs: FinalizationStatus) -> bool {
-        FinalizationStatusIntoU8::into(lhs) == rhs.into()
+impl PartialEqStrategy of PartialEq<Strategy> {
+    fn eq(lhs: @Strategy, rhs: @Strategy) -> bool {
+        lhs.address == rhs.address
+            && poseidon::poseidon_hash_span(
+                lhs.params.span()
+            ) == poseidon::poseidon_hash_span(rhs.params.span())
     }
 
-    fn ne(lhs: FinalizationStatus, rhs: FinalizationStatus) -> bool {
-        FinalizationStatusIntoU8::into(lhs) != rhs.into()
+    fn ne(lhs: @Strategy, rhs: @Strategy) -> bool {
+        !(lhs.clone() == rhs.clone())
```

Comment:
Sucks we need to clone here... I mean in theory we're just checking stuff, not modifying, so I'm guessing there should be a way of not cloning, right?

Reply:
not sure. the impls in the standard lib dereference the snapshot: https://github.com/starkware-libs/cairo/blob/22ccda9eb66afcb6465e4778ddc310a5d09e3c91/corelib/src/integer.cairo#L297C9-L297C9 so doesnt look like there is a way to compare snapshots directly

Reply:
So for `u8` yeah sure but for `param` it seems very expensive to do that.. but yeah maybe it's my `Rust` background talking here... :p 

But like the logic is: if you're comparing values, you def don't need to take ownership of them :p